### PR TITLE
Koopman Lifting for Tandem Interference: linear operator in lifted space

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Learn a **Koopman operator** that lifts the nonlinear tandem inter-foil interference dynamics to a higher-dimensional linear space. In this lifted space, the fore-foil's effect on the aft-foil is approximately linear: `z_aft_corrected = z_aft + K_tandem @ z_fore`, where K_tandem is a learned low-rank linear operator. This explicitly disentangles the inter-foil coupling as a linear map in the lifted representation.

**Why this helps p_tan:** The tandem inter-foil interference is dominated by downwash (fore-foil wake changes the aft-foil effective AoA). This is approximately linear in the lifted space of boundary layer parameters. By learning K_tandem as a low-rank matrix, the model can generalize the tandem interference pattern to unseen gap/stagger combinations — directly targeting p_tan. The Koopman lifting is applied ONLY to tandem samples (identity for single-foil), preserving p_in.

**Literature:**
- "Invertible Koopman neural operator for data-driven modeling of PDEs" (Papers With Code, Mar 2025)
- "Stochastic reduced-order Koopman model for turbulent flows" (arXiv:2503.22649, Mar 2025)
- "CGKN: A Deep Learning Framework for Modeling Complex Dynamical Systems" (arXiv:2410.20072)

### Architecture

Add a **Koopman Tandem Lifting** module between the Transolver backbone output and the SRF input. The module has three components:

1. **Lifting Encoder** φ: Maps backbone hidden states to a lifted Koopman space.
   - Input: backbone hidden state h ∈ R^192 per node
   - Architecture: `Linear(192, 256) → GELU → Linear(256, 256)`
   - Output: lifted state z ∈ R^256

2. **Koopman Operator** K_tandem: A learnable low-rank linear map for tandem correction.
   - Parameterize as: `K = U @ diag(sigma) @ V^T` where U ∈ R^(256×32), V ∈ R^(256×32), sigma ∈ R^32
   - This is a rank-32 approximation — keeps parameter count low (~16K params)
   - For each aft-foil node: `z_aft_corrected = z_aft + K @ mean_pool(z_fore)` where mean_pool(z_fore) is the mean of all fore-foil nodes' lifted states
   - For fore-foil nodes: `z_fore_out = z_fore` (identity — no correction needed)
   - For single-foil: skip the Koopman operator entirely (identity path)

3. **Projection Decoder** ψ: Maps back from lifted space to backbone dimension.
   - Architecture: `Linear(256, 192) → GELU → Linear(192, 192)`
   - Add residual: `h_out = h + ψ(z_corrected)` (residual connection preserves backbone information)

4. **Reconstruction Loss** (auxiliary): Enforce invertibility of the lifting.
   - L_recon = ||ψ(φ(h)) - h||₁ with weight 0.05
   - This prevents the lifting from collapsing to a trivial solution

### Implementation Details

Add flags:
- `--koopman_tandem`: activate the Koopman lifting module
- `--koopman_rank 32`: rank of K_tandem (default 32)
- `--koopman_dim 256`: dimension of lifted space (default 256)
- `--koopman_recon_weight 0.05`: reconstruction loss weight

Key implementation points:
- The Koopman module goes AFTER the last TransolverBlock and BEFORE the SRF
- For tandem samples, identify fore vs aft foil nodes using the `foil_role` or `foil_id` batch field
- `mean_pool(z_fore)` averages all fore-foil nodes' lifted states into a single 256-dim vector, then K maps this to a correction for each aft-foil node
- The reconstruction loss is added to the total loss (integrated with PCGrad if applicable — use the pressure task head)
- Use standard Xavier initialization for U, V; initialize sigma as ones (K starts as a random rank-32 matrix)

### Training Setup

```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --koopman_tandem --koopman_rank 32 --koopman_dim 256 --koopman_recon_weight 0.05 \
  --wandb_group koopman-tandem \
  --seed 42
```

Run 2 seeds (42, 73) as standard.

### What to Watch

- p_tan is the primary target — the Koopman correction specifically targets aft-foil predictions
- p_in should NOT change (Koopman is identity for single-foil)
- Monitor `koopman_recon_loss` in W&B — if it doesn't decrease, the lifting is degenerate
- Monitor singular values of K (log sigma) — if they collapse to zero, the operator isn't learning
- Epoch time overhead should be minimal (<5%) — the Koopman module is lightweight

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

Baseline PR: #2357 (Vortex-Panel Induced Velocity)

Reproduce baseline:
```
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2380 because ADVISOR: Closing — Koopman tandem lifting is a dead end. EMA metrics (2-seed avg):
- p_in: 14.09 (baseline 11.872, +18.7%)
- p_oodc: 8.09 (baseline 7.459, +8.4%)
- p_tan: 30.18 (baseline 26.319, +14.7%)
- p_re: 6.69 (baseline 6.229, +7.4%)

W&B: ik7k6o9l (s42), 4col5vx2 (s73). All metrics significantly worse. The linear Koopman operator in the lifted space could not capture the nonlinear tandem interference dynamics better than the standard Transolver. The lifting transformation likely added noise without providing useful structure. Moving you to a new experiment..

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Metrics vs Baseline

| Metric | Baseline | Seed 42 | Seed 73 | **2-seed avg** | Δ |
|--------|----------|---------|---------|----------------|---|
| **p_in** | 11.872 | 14.750 | 13.432 | **14.091** | +2.22 (+18.7%) ❌ |
| p_oodc | 7.459 | 8.180 | 7.992 | **8.086** | +0.63 (+8.4%) ❌ |
| **p_tan** | 26.319 | 30.079 | 30.278 | **30.179** | +3.86 (+14.7%) ❌ |
| **p_re** | 6.229 | 6.684 | 6.701 | **6.692** | +0.46 (+7.4%) ❌ |

All metrics degraded vs baseline. Koopman tandem lifting did not help.

### Val loss summary (best EMA)

| Split | Seed 42 | Seed 73 | Avg |
|-------|---------|---------|-----|
| val_in_dist | 0.447 | 0.408 | 0.427 |
| val_ood_cond | 0.474 | 0.469 | 0.472 |
| val_ood_re | 0.437 | 0.438 | 0.437 |
| val_tandem | 1.255 | 1.253 | 1.254 |

### Run details
- Seed 42 W&B:  (nezuko/koopman-tandem-s42)
- Seed 73 W&B:  (nezuko/koopman-tandem-s73)
- Group: 
- Koopman params: 218,016 (dim=256, rank=32)
- Training epochs: ~135-145 (3h timeout)

### Commands used
```
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent nezuko --wandb_name "nezuko/koopman-tandem-s42" --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --koopman_tandem --koopman_rank 32 --koopman_dim 256 --koopman_recon_weight 0.05 --wandb_group koopman-tandem --seed 42

# Seed 73 (same with --seed 73 on CUDA_VISIBLE_DEVICES=1)
```

## Research

The researcher-agent confirmed:
- Koopman operator literature (FlowDMD, CF-INN, CGKN) uses linearity + reconstruction losses — our design follows this
- Low-rank parameterization is standard and appropriate
- Mean-pooling fore-foil nodes is a reasonable aggregation choice
- Invertibility enforcement via reconstruction loss prevents trivial collapse

## What happened

**Negative result.** The Koopman lifting degraded all metrics across the board, most importantly p_tan went from 26.32 → 30.18 (+14.7%).

**Why it likely failed:**

1. **Architectural positioning issue**: The Koopman module transforms  (the pre-last-block backbone features, ~192-dim). These corrected hidden states only feed into the SRF (Surface Refinement Head). They do NOT affect the main backbone predictions (), which come from the last TransolverBlock applied to the original un-corrected . So the Koopman correction's only pathway to affect predictions is through the SRF, which is itself a small auxiliary MLP. Both Koopman and SRF learn simultaneously from scratch, creating a harder joint optimization.

2. **Two-module cold-start problem**: The SRF was previously trained to correct the backbone's original hidden states. Adding Koopman creates a new intermediate representation that the SRF must also learn from scratch. Instead of one cold-start problem (SRF alone), we now have two entangled cold-starts.

3. **The Koopman correction didn't learn useful signal**: The reconstruction loss (ensuring ψ(φ(h)) ≈ h) enforces auto-encoding, but doesn't ensure the correction K @ mean_pool(z_fore) actually encodes tandem interference. The sigma values in W&B showed they were learning (non-zero), but the correction might have learned noise rather than the fore-foil → aft-foil interference signal.

4. **p_in also degraded**: If p_in (single-foil) is worse, the Koopman module harmed the single-foil path too — possibly because SRF is now distracted from its original purpose by having to handle Koopman-modified hidden states even for single-foil samples where Koopman applies identity.

5. **218K extra parameters** added noise to the optimization, competing with the existing well-tuned baseline.

## Suggested follow-ups

1. **Insert Koopman INSIDE the Transolver backbone** (between  output and  input), so it directly affects the backbone's final predictions — not just the SRF input. This is the architecturally correct placement per the PR spec and would have maximum impact.

2. **Freeze the backbone, fine-tune Koopman+SRF** for a few epochs before full joint training — to avoid the cold-start entanglement.

3. **Koopman as direct prediction correction** (like SRF, but only for aft-foil nodes): skip the reconstruction path, just apply a low-rank update directly to the predicted pressure for aft-foil nodes, computed from fore-foil node pressure statistics.

4. **Smaller Koopman** (rank=8, dim=128) with higher recon weight (0.1) to reduce parameter overhead and strengthen the invertibility constraint.

5. **Explicit linearity loss**: Add a loss term ||z_aft_next - K @ z_fore||₁ where z_aft_next is from the next batch's aft features — this would more directly train K to be the Koopman operator rather than relying on indirect signal from predictions.
